File: app/src/main/java/org/mifos/mobile/ui/adapters/LoanRepaymentScheduleAdapter.java
package org.mifos.mobile.ui.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.evrencoskun.tableview.adapter.AbstractTableAdapter;
import com.evrencoskun.tableview.adapter.recyclerview.holder.AbstractViewHolder;

import org.mifos.mobile.R;
import org.mifos.mobile.injection.ApplicationContext;
import org.mifos.mobile.models.accounts.loan.Periods;
import org.mifos.mobile.models.accounts.loan.tableview.Cell;
import org.mifos.mobile.models.accounts.loan.tableview.ColumnHeader;
import org.mifos.mobile.models.accounts.loan.tableview.RowHeader;
import org.mifos.mobile.utils.CurrencyUtil;
import org.mifos.mobile.utils.DateHelper;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Rajan Maurya on 04/03/17.
 */

public class LoanRepaymentScheduleAdapter extends
        AbstractTableAdapter<ColumnHeader, RowHeader, Cell> {

    private String currency = "";

    public void setCurrency(String currency) {
        this.currency = currency;
    }

    @Inject
    LoanRepaymentScheduleAdapter(@ApplicationContext Context context) {
        super(context);
    }

    class CellViewHolder extends AbstractViewHolder {

        @BindView(R.id.cell_data)
        TextView tvCell;

        CellViewHolder(View v) {
            super(v);
            ButterKnife.bind(this, v);
        }
    }

    @Override
    public AbstractViewHolder onCreateCellViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.cell_loan_repayment_schedule, parent, false);
        return new CellViewHolder(view);
    }

    @Override
    public void onBindCellViewHolder(AbstractViewHolder holder, Object cellItemModel,
                                     int columnPosition, int rowPosition) {
        Cell cell = (Cell) cellItemModel;
        Periods period = (Periods) cell.getData();

        // Get the holder to update cell item text
        CellViewHolder viewHolder = (CellViewHolder) holder;

        switch (columnPosition) {
            case 0:
                viewHolder.tvCell.setText(DateHelper.getDateAsString(period.getDueDate()));
                break;
            case 1:
                viewHolder.tvCell.setText(mContext.getString(R.string.string_and_double,
                        currency, period.getPrincipalOriginalDue()));
                break;
            case 2:
                viewHolder.tvCell.setText(mContext.getString(R.string.string_and_string,
                        currency, CurrencyUtil.formatCurrency(mContext, period.
                                getPrincipalLoanBalanceOutstanding())));
                break;
            default:
                viewHolder.tvCell.setText("");
        }
    }

    class ColumnHeaderViewHolder extends AbstractViewHolder {

        @BindView(R.id.column_header_textView)
        TextView tvColumnHeader;

        ColumnHeaderViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    @Override
    public AbstractViewHolder onCreateColumnHeaderViewHolder(ViewGroup parent, int viewType) {

        // Get Column Header xml Layout
        View layout = LayoutInflater.from(mContext).inflate(R.layout
                .column_header_loan_repayment_schedule, parent, false);

        // Create a ColumnHeader ViewHolder
        return new ColumnHeaderViewHolder(layout);
    }

    @Override
    public void onBindColumnHeaderViewHolder(AbstractViewHolder holder,
                                             Object columnHeaderItemModel, int position) {
        ColumnHeader columnHeader = (ColumnHeader) columnHeaderItemModel;

        // Get the holder to update cell item text
        ColumnHeaderViewHolder columnHeaderViewHolder = (ColumnHeaderViewHolder) holder;

        columnHeaderViewHolder.tvColumnHeader.setText(String.valueOf(columnHeader.getData()));
    }

    class RowHeaderViewHolder extends AbstractViewHolder {

        @BindView(R.id.row_header_textview)
        TextView tvRowHeader;

        RowHeaderViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    @Override
    public AbstractViewHolder onCreateRowHeaderViewHolder(ViewGroup parent, int viewType) {

//         Get Row Header xml Layout
        View layout = LayoutInflater.from(mContext).inflate(R.layout
                .row_header_loan_repayment_schedule, parent, false);

        // Create a Row Header ViewHolder
        return new RowHeaderViewHolder(layout);
//        return null;
    }

    @Override
    public void onBindRowHeaderViewHolder(AbstractViewHolder holder, Object rowHeaderItemModel, int
            position) {
        RowHeader rowHeader = (RowHeader) rowHeaderItemModel;

        // Get the holder to update row header item text
        RowHeaderViewHolder rowHeaderViewHolder = (RowHeaderViewHolder) holder;
        rowHeaderViewHolder.tvRowHeader.setText(String.valueOf(rowHeader.getData()));
    }

    @Override
    public View onCreateCornerView() {
        return LayoutInflater.from(mContext).inflate(R.layout
                .corner_view_loan_repayment_schedule, null, false);
    }

    @Override
    public int getColumnHeaderItemViewType(int columnPosition) {
        // The unique ID for this type of column header item
        // If you have different items for Cell View by X (Column) position,
        // then you should fill this method to be able create different
        // type of CellViewHolder on "onCreateCellViewHolder"
        return 0;
    }

    @Override
    public int getRowHeaderItemViewType(int rowPosition) {
        // The unique ID for this type of row header item
        // If you have different items for Row Header View by Y (Row) position,
        // then you should fill this method to be able create different
        // type of RowHeaderViewHolder on "onCreateRowHeaderViewHolder"
        return 0;
    }

    @Override
    public int getCellItemViewType(int columnPosition) {
        // The unique ID for this type of cell item
        // If you have different items for Cell View by X (Column) position,
        // then you should fill this method to be able create different
        // type of CellViewHolder on "onCreateCellViewHolder"
        return 0;
    }

}

File: app/src/main/java/org/mifos/mobile/ui/adapters/LoanRepaymentScheduleAdapter.kt
package org.mifos.mobile.ui.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.evrencoskun.tableview.adapter.AbstractTableAdapter
import com.evrencoskun.tableview.adapter.recyclerview.holder.AbstractViewHolder
import org.mifos.mobile.R
import org.mifos.mobile.core.model.entity.accounts.loan.Periods
import org.mifos.mobile.core.model.entity.accounts.loan.tableview.Cell
import org.mifos.mobile.core.model.entity.accounts.loan.tableview.ColumnHeader
import org.mifos.mobile.core.model.entity.accounts.loan.tableview.RowHeader
import org.mifos.mobile.databinding.CellLoanRepaymentScheduleBinding
import org.mifos.mobile.databinding.ColumnHeaderLoanRepaymentScheduleBinding
import org.mifos.mobile.databinding.CornerViewLoanRepaymentScheduleBinding
import org.mifos.mobile.databinding.RowHeaderLoanRepaymentScheduleBinding
import org.mifos.mobile.utils.CurrencyUtil.formatCurrency
import org.mifos.mobile.utils.DateHelper.getDateAsString
import javax.inject.Inject

/**
 * Created by Rajan Maurya on 04/03/17.
 */
class LoanRepaymentScheduleAdapter @Inject internal constructor() :
    AbstractTableAdapter<ColumnHeader?, RowHeader?, Cell?>() {

    private var currency: String? = ""
    private var columnWidth: Double = 0.0
    fun setCurrency(currency: String?) {
        this.currency = currency
    }

    internal inner class CellViewHolder(private val bindingCellView: CellLoanRepaymentScheduleBinding) :
        AbstractViewHolder(bindingCellView.root) {
        fun bind(columnPosition: Int, period: Periods) {
            with(bindingCellView) {
                var principal: Double?
                when (columnPosition) {
                    0 -> cellData.text = getDateAsString(period.dueDate)
                    1 -> {
                        principal = period.principalOriginalDue
                        if (principal == null) {
                            principal = 0.00
                        }
                        cellData.text = itemView.context.getString(
                            R.string.string_and_double,
                            currency,
                            principal,
                        )
                    }

                    2 -> {
                        principal = period.principalLoanBalanceOutstanding
                        if (principal == null) {
                            principal = 0.00
                        }
                        cellData.text = itemView.context.getString(
                            R.string.string_and_string,
                            currency,
                            formatCurrency(itemView.context, principal),
                        )
                    }

                    else -> cellData.text = ""
                }
                cellContainer.layoutParams?.width = columnWidth.toInt()
                cellData.requestLayout()
            }
        }
    }

    override fun onCreateCellViewHolder(parent: ViewGroup, viewType: Int): AbstractViewHolder {
        val bindingCellView = CellLoanRepaymentScheduleBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false,
        )
        return CellViewHolder(bindingCellView)
    }

    override fun onBindCellViewHolder(
        holder: AbstractViewHolder,
        cellItemModel: Cell?,
        columnPosition: Int,
        rowPosition: Int,
    ) {
        val (data) = cellItemModel as Cell
        val period = data as Periods

        holder as CellViewHolder
        holder.bind(columnPosition, period)
    }

    internal inner class ColumnHeaderViewHolder(private val bindingColumnHeader: ColumnHeaderLoanRepaymentScheduleBinding) :
        AbstractViewHolder(bindingColumnHeader.root) {
        fun bind(columnHeaderText: String) {
            with(bindingColumnHeader) {
                columnHeaderTextView.text = columnHeaderText
                columnHeaderContainer.layoutParams?.width = columnWidth.toInt()
                columnHeaderTextView.requestLayout()
            }
        }
    }

    override fun onCreateColumnHeaderViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): AbstractViewHolder {
        val bindingColumnHeader = ColumnHeaderLoanRepaymentScheduleBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false,
        )
        return ColumnHeaderViewHolder(bindingColumnHeader)
    }

    override fun onBindColumnHeaderViewHolder(
        holder: AbstractViewHolder,
        columnHeaderItemModel: ColumnHeader?,
        position: Int,
    ) {
        val (data) = columnHeaderItemModel as ColumnHeader

        holder as ColumnHeaderViewHolder
        holder.bind(data.toString())
    }

    internal inner class RowHeaderViewHolder(private val bindingRowHeader: RowHeaderLoanRepaymentScheduleBinding) :
        AbstractViewHolder(bindingRowHeader.root) {
        fun bind(rowHeaderText: String) {
            bindingRowHeader.rowHeaderTextview.text = rowHeaderText
        }
    }

    override fun onCreateRowHeaderViewHolder(parent: ViewGroup, viewType: Int): AbstractViewHolder {
        val bindingRowHeader = RowHeaderLoanRepaymentScheduleBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false,
        )
        return RowHeaderViewHolder(bindingRowHeader)
    }

    override fun onBindRowHeaderViewHolder(
        holder: AbstractViewHolder,
        rowHeaderItemModel: RowHeader?,
        position: Int,
    ) {
        val (data) = rowHeaderItemModel as RowHeader
        holder as RowHeaderViewHolder
        holder.bind(data.toString())
    }

    override fun onCreateCornerView(parent: ViewGroup): View {
        val bindingRowHeader = CornerViewLoanRepaymentScheduleBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false,
        )
        return bindingRowHeader.root
    }

    override fun getColumnHeaderItemViewType(columnPosition: Int): Int {
        // The unique ID for this type of column header item
        // If you have different items for Cell View by X (Column) position,
        // then you should fill this method to be able create different
        // type of CellViewHolder on "onCreateCellViewHolder"
        return 0
    }

    override fun getRowHeaderItemViewType(rowPosition: Int): Int {
        // The unique ID for this type of row header item
        // If you have different items for Row Header View by Y (Row) position,
        // then you should fill this method to be able create different
        // type of RowHeaderViewHolder on "onCreateRowHeaderViewHolder"
        return 0
    }

    override fun getCellItemViewType(columnPosition: Int): Int {
        // The unique ID for this type of cell item
        // If you have different items for Cell View by X (Column) position,
        // then you should fill this method to be able create different
        // type of CellViewHolder on "onCreateCellViewHolder"
        return 0
    }

    fun setColumnWidth(columnWidth: Double) {
        this.columnWidth = columnWidth
    }
}
